helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -200,84 +199,103 @@ private static void retainExistingReplicas(Map<String, Set<AssignableReplica>> r
             stateInstanceMap.getOrDefault(replica.getPartitionName(), Collections.emptyMap())
                 .getOrDefault(replica.getReplicaState(), Collections.emptySet());
         if (validInstances.isEmpty()) {
-          // Removing by comparing with the baseline assignment.
+          // Removing the replica if it is not known in the assignment map.
           replicaIter.remove();
         } else {
-          // Remove the instance from the state map record, so it won't be picked up again for
-          // the other replica checkup.
+          // Remove the instance from the state map record after processing so it won't be
+          // double-processed as we loop through all replica
           validInstances.remove(validInstances.iterator().next());
         }
       }
     });
   }
 
   /**
-   * Find the minimum set of replicas that need to be reassigned by comparing the Best
-   * possible assignment with the Baseline assignment.
-   * A replica needs to be reassigned if either of the following conditions is true:
-   * 1. The partition allocation in the Baseline and the Best possible assignment are different.
-   * And the allocation in the Baseline is valid. So it is worthwhile to move it.
-   * 2. The partition allocation is not in the Baseline or the Best possible assignment.
-   * Otherwise, the rebalancer just keeps the current Best possible assignment allocation.
+   * Find the minimum set of replicas that need to be reassigned by comparing the current assignment
+   * with the ideal assignment.
+   * A replica needs to be reassigned or newly assigned if either of the following conditions is true:
+   * 1. The partition allocation in the ideal assignment and the current assignment are different.
+   * And the allocation in the ideal assignment is valid. So it is worthwhile to move it.
+   * 2. The partition allocation is not in the ideal assignment nor the current assignment. Or those
```

Comment:
The partition allocation **(the instance the replica is placed on)** is in **neither** the ideal assignment nor the current assignment.